ios/engine/KMEI/KeymanEngine/Classes/Resource Data/FontManager.swift

```diff
@@ -133,32 +168,54 @@ public class FontManager {
 
     return font.isRegistered
   }
-
-  public func registerFonts(in directory: URL) {
+  
+  private func listFonts(in directory: URL) -> [URL] {
     guard let urls = try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil) else {
       let message = "Could not list contents of directory \(directory)"
       os_log("%{public}s", log:KeymanEngineLogger.resources, type: .error, message)
-      return
-    }
-    for url in urls where url.lastPathComponent.hasFontExtension {
-      _ = registerFont(at: url)
+      return []
     }
+    return urls.filter { $0.lastPathComponent.hasFontExtension }
+  }
+
+  public func registerFonts(in directory: URL) {
+    let fontsToRegister = listFonts(in: directory)
+    registerListedFonts(Set(fontsToRegister))
   }
 
   public func unregisterFonts(in directory: URL, fromSystemOnly: Bool = true) {
-    guard let urls = try? FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil) else {
-      let message = "Could not list contents of directory \(directory)"
-      os_log("%{public}s", log:KeymanEngineLogger.resources, type: .error, message)
-      return
-    }
-    for url in urls where url.lastPathComponent.hasFontExtension {
+    let fontsToUnregister = listFonts(in: directory)
+    for url in fontsToUnregister {
       _ = unregisterFont(at: url, fromSystemOnly: fromSystemOnly)
     }
   }
 
-  private func fontExists(_ fontName: String) -> Bool {
-    return UIFont.familyNames.contains { familyName in
-      UIFont.fontNames(forFamilyName: familyName).contains(fontName)
+  /**
+   * Queries the system for existing registrations for the specified fonts with a single batch run.
+   * Only fonts that could not be found will be returned within the result set.
+   */
+  private func missingFonts(from fontNames: Set<String>) -> Set<String> {
+    // Arrays are treated 'by value'; it's already a shallow copy.
```

Comment:
This comment doesn't belong -- it's about how Swift works, not about how our code works

```suggestion
```